File: leetcode_java/java/com/goinhn/thorwcode/leetcode/question/_1401_1500/Sol1470.java
package com.goinhn.thorwcode.leetcode.question._1401_1500;

/**
 * 1470.重新排列数组
 *
 * @author goinhn
 * @date 2022/8/29
 */
public class Sol1470 {

    /**
     * time:O(n)
     * space:O(n)
     */
    private static class Solution1 {
        public int[] shuffle(int[] nums, int n) {
            int[] res = new int[nums.length];
            int left = 0;
            int right = n;
            int idx = 0;
            while (idx < res.length) {
                res[idx++] = nums[left++];
                res[idx++] = nums[right++];
            }

            return res;
        }
    }

    /**
     * 用多余的高位来存储排好序之后的数字，最后将高位的数字回正即可
     * time:O(n)
     * space:O(1)
     */
    private static class Solution2 {
        public int[] shuffle(int[] nums, int n) {
            for (int i = 0; i < nums.length; i++) {
                int j = i < n ? i * 2 : (i - n) * 2 + 1;
                nums[j] |= (nums[i] & 1023) << 10;
            }

            for (int i = 0; i < nums.length; i++) {
                nums[i] >>= 10;
            }

            return nums;
        }
    }

}

File: leetcode_go/question/_1401_1500/_1470/sol_1470.go
// Package _1470 重新排列数组
// @author goinhn
// @date 2022/8/29
package _1470

// time:O(n)
// space:O(n)
func shuffle1(nums []int, n int) []int {
	res := make([]int, len(nums))
	left := 0
	right := n
	idx := 0
	for idx < len(nums) {
		res[idx] = nums[left]
		idx++
		left++
		res[idx] = nums[right]
		idx++
		right++
	}

	return res
}

// 通过多余的位来存储排序好的结果
// time:O(n)
// space:O(1)
func shuffle2(nums []int, n int) []int {
	for i := range nums {
		var j int
		if i < n {
			j = 2 * i
		} else {
			j = 2*(i-n) + 1
		}
		nums[j] |= (nums[i] & 1023) << 10
	}

	for i := range nums {
		nums[i] >>= 10
	}

	return nums
}
